Add Melodot proposal

DarkingLee opened the pull request:
#### Project Abstract

Melodot is an incentive-compatible data availability layer. Melodot introduces the role of farmers by combining PoSpace, alleviating the system's dependence on the minimum number of honest nodes assumption, and completing an incentive-compatible distributed generation and data storage scheme. Consensus validators now act more like light clients, improving future scalability.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (bank details via email _or_ BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (USDT) address in the application).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

DarkingLee commented:
Hey @Noc2 and @keeganquigley , I hope you're both doing well! I created the proposal and would love to get your suggestions!

Noc2 reviewed (changes requested):
Thanks for the application, and sorry for the late reply here. I have some quick initial feedback: 
- All the headlines of your milestones currently say: "Example — Basic functionality". Could you update this? 
- The estimation of the FTE and total costs seem strange to me. For example, milestone 4 has an FTE of 2 and one month compared to milestone 2, which has only an FTE of 1 and the same total costs. 
- Are you aware of avail: https://github.com/availproject?

DarkingLee commented:
hey @Noc2 , Thank you for your reply.

- I have updated them.
- It indeed seemed strange, so I carefully reviewed the milestones and made the following adjustments:
    - Milestone 1 now uses the month format for better clarity.
    - Based on the actual situation, I re-evaluated the duration of Milestone 2 and adjusted it to 1.5 months.
    - I have aligned for total costs / FTE * estimated duration, and now they are approximately consistent.
    - The total costs have been reduced by 2,000 compared to before.
- Yes, I am aware of Avail. Melodot is quite different from it in terms of data arrangement, distributed generation, and handling of the "honest light clients".

Noc2 reviewed (changes requested):
Thanks for the quick reply. I will mark the application already as ready for review. But I need to spend a little bit more time looking into it. One question that I currently have is: do you think Melodot could potentially be a parachain, and what issues might you face in this regard?

keeganquigley reviewed (commented):
Thanks for the application @DarkingLee, your whitepaper link is asking me to log into a notion account. Is there a public link where I can read it?

keeganquigley commented on applications/Melodot.md:
```diff
@@ -0,0 +1,238 @@
+| **0b.** | Documentation                  | We will provide both inline documentation of the code and a basic tutorial that explains how a user can spin up a client, connect to the client management application and data through a browser, and create a local development chain. |
+| **0c.** | Testing and Testing Guide      | Higher-level integration tests and unit tests for all modules. In the guide, we will describe how to run these tests. |
+| **0d.** | Docker                         | We will provide Dockerfiles to start multiple nodes with different roles, create a local test network, and run all integration tests. |
+| 1.      | Substrate pallet: excutive_das | Depends on the `frame-executive` pallet, extending the DAS-related scheduling and data structures. |
+| 2.      | Substrate pallet: system_das   | Depends on the `frame-system` pallet, extending the DAS-related actual operations. |
```
The dependents are listed here, but could you expand this to include what functions would be built as part of the pallet?

DarkingLee replied:
OK, it has been updated.

keeganquigley commented on applications/Melodot.md:
```diff
@@ -0,0 +1,238 @@
+
+### Similar Projects
+
+There are currently several data availability layer projects, including Ethereum Danksharding, Celestia, Avail, and Eigen DA. Our main difference from them is the introduction of farmers, which solves many tricky problems faced by the data availability layer. Unlike Danksharding, we decouple an independent data availability layer, which is the same principle as Polkadot not supporting smart contr ...
```
It also reminds me of the old proof-of-capacity project [Signum (formerly Burstcoin)](https://signum.network/), combined with Chia's proof-of-space. Can you explain who would be the primary users of this project once it is complete?

DarkingLee replied:
Yes, there are various implementations of proof-of-space / proof-of-capacity. The difference lies in the fact that Melodot utilizes them to store availability data,  so the primary users of Melodot are rollup applications.

keeganquigley commented on applications/Melodot.md:
```diff
@@ -0,0 +1,238 @@
+| **0d.** | Docker                         | We will provide Dockerfiles to start multiple nodes with different roles, create a local test network, and run all integration tests. |
+| 1.      | Substrate pallet: excutive_das | Depends on the `frame-executive` pallet, extending the DAS-related scheduling and data structures. |
+| 2.      | Substrate pallet: system_das   | Depends on the `frame-system` pallet, extending the DAS-related actual operations. |
+| 3.      | Substrate pallet: melo_store   | Pallet for managing apps and submitting data.                |
```
Can you elaborate on how this pallet manages apps and submits the validity data?

DarkingLee replied:
OK, it has been updated.

DarkingLee commented:
> Thanks for the quick reply and update. I will mark the application already as ready for review. But I need to spend a little bit more time looking into it. One question that I currently have is: do you think Melodot could potentially be a parachain, and what issues might you face in this regard?感谢您的快速回复和更新。我会将申请标记为已准备好接受审核。但我需要花更多的时间来研究它。我目前的一个问题是：您认为 Melodot 是否有可能成为一条平行链，在这方面您可能会面临哪些问题？

Melodot is not well-suited to directly function as a parachain, but rather as a parachain through a "data availability bridge", essentially because as a decoupled data availability layer:

- It operates on a different security model. For instance, on the consensus layer, GRANDPA reaches agreements on chains rather than blocks, which is useful in extreme cases to ensure fast and secure finality, but also has the potential to bring multiple blocks to finality at the same time. The data availability layer, on the other hand, expects a uniform finality, which would otherwise introduce uncertain transaction delays for rollup transactions in the settlement layer. In the long run, Melodot needs to have a more suitable finality gadget. In addition, the data availability layer requires many changes at the network layer (which is still an open research area). These determine that Melodot is not well-suited to directly function as a parachain.
- The purpose of the data availability layer is to eliminate the reliance on the "honest majority" to secure data availability even in the case of a 51% attack, so I think it is a "Muda" to use it directly as a parachain.

As Melodot matured, we used the Data Availability Bridge as a parachain. It was designed as a neutral bridge (allowing other data availability layers), and other parachains acted as settlement layers, getting data availability guarantees from Melodot parallel chains and "secure interoperability of settlement layers" from the security of polkadot.

DarkingLee commented:
> Thanks for the application @DarkingLee, your whitepaper link is asking me to log into a notion account. Is there a public link where I can read it?

I reset the link and now it doesn't need the login anymore.

DarkingLee commented:
@keeganquigley Thank you for your feedback, sorry for the late reply.

- Chia-style proof-of-space requires an initialization process called plots. This process writes data to the user's storage device and consumes significant resources. However, after this initialization, the resource consumption is minimal, specifically in the case of Melodot, it is less than that of a full node. The criticism of Chia, in my opinion, primarily revolves around filling users' disks with useless random data.
- I have been focusing on the development of social networks and reputation systems (implemented through Substrate), and in this process, I have identified some fundamental problems that have yet to be solved. For example, transaction ordering is not crucial in social networks, but in DeFi, significant costs are incurred to achieve fair transaction ordering. Under blockchain technology, they are given the same level of security. This is not an issue that can be simply solved by scaling alone. To address this, we conducted extensive research, even rejecting several investors. After multiple design attempts, we decided to adopt Rollup technology to solve this problem. I aim to decouple the data availability layer and sequencer in the Polkadot ecosystem through parachain, where Polkadot's high security provides the settlement layer with security and scalability. Of course, further research and implementation are still required, however, I hope that at least through the data availability layer, bridging parallel chains with DA, and the SDK, we can bring another vast possibility space to the Polkadot ecosystem.
- We have a semi-annual plan, and once this grant is complete, we will initiate a community through a DAO, leveraging the DAO's economic system to drive project collaboration and obtain a small amount of initial funding. Afterward, we will secure early reliable funding through a method called *Quadratic Launch*, which combines elements of an IDO and quadratic funding. The initial tokens will be allocated to the donors of quadratic funding. We have a [detailed plan](https://zerodao.notion.site/Phased-development-strategy-management-1615ff6679134503b9426e5e99d3c53f?pvs=4) that includes different stages, financial planning, product roadmap, and more.

Thanks!

Noc2 reviewed (approved):
I'm sorry for not getting back to you sooner. To be honest, my biggest concern here and with the application is that you are trying to do too many things at the moment. But I'm happy to approve it since some of it might definitely be useful for other projects.

DarkingLee commented:
@Noc2, Thank you for the review, and I understand that you must have been quite busy lately. Your feedback is valuable to me, and I will definitely take note of your concern about trying to do too many things at the moment.

takahser reviewed (approved):
@DarkingLee thanks, you've clearly put a lot of effort into the whitepaper and the proposal - it's well-written and contains interesting aspects, both from an economical and game-theoretical as well as technical point of view. It's the first time I've seen the concept of price elasticity applied to a web3 project, I've never thought of that before but in retrospect it seems obvious:) Anyway, happy to approve here as well.

burdges commented:
> The current data availability layer scheme requires the assumption that the network has at least the minimum number of honest nodes.

This statement is true.

> However, due to the need to prevent data retention attacks, the samplers are required to be anonymous, making it difficult to measure the number of samplers. 

Polkadot does not require separate samplers for availability.  The validators all "sample" exactly the one piece they hold themselves.  Intersection of 2/3rd honest and 2/3rd claiming availability, says 1/3rd really are available, which suffices for reconstruction. 

In other words, 1d RS has a better undecodable ratio, aka adversary must hide 2/3rd to hide anything, while a 2d RS permits hiding a small chunk of the block if the adversary hides only 25%, which further increases sampling requirements.

We do "sample" for validity in the approvals protocol, which includes proving correctness of the available data, but that's an separate concern.  We do assume roughly a synchronous gossip protocol for approval assignment announcements though, but we do not require anonymity when sampling for validity, thanks to the no show system.  

> At the same time, samplers are more concerned with data related to themselves, resulting in an uneven distribution of total active samplers over time. 

Ain't clear what you mean here, but it's evenly distributed among the nodes who claim to hold their chunk.

Aside from an undecodable ratio meshing cleanly with our security assumptions elsewhere, we abandoned wider sampling ideas used by Celestia like fishmermen or watchtowers because we concluded there was no mechanism for incentivizing correct behavior.  We know validators can be incentivized, although this turns out nastier than one likes. 

I've not looked into the rest but this paragraph jumped out at me, so figured I'd comment here.  Carry on  :)

DarkingLee commented:
@burdges Thank you for your comment.

1. I greatly appreciate you sharing Polkadot's design, it indeed sounds intriguing. Polkadot's solution certainly seems elegant. However, there are fundamental distinctions between Melodot and Polkadot. Melodot serves as an independent DA layer, designed to cater to the settlement layer and rollup applications. The "settlement layer" in itself is another chain that's loosely coupled with the DA layer, and its security does not rely on 2/3 of honest actors from the DA layer. This presents additional design challenges for Melodot.

> At the same time, samplers are more concerned with data related to themselves, resulting in an uneven distribution of total active samplers over time.
2. The crux of the matter here is that we don't rely on validators having data sufficient for decoding. Instead, we depend on a significant number of light clients randomly sampling across the network. This strategy has the advantage of considerably increasing the data payload of each block but requires proportionately more sampling nodes to support. The design can be challenged when the number of samplers for each block becomes inconsistent. For Celestia, I believe they might be facing similar design decisions. And Polkadot seems to have no such worries :)
